### DISCUSSION TITLE
How to set Client using GSSAPI NTLM?

### DISCUSSION THREAD

**Discussion by ZinedineR** [wneessen/go-mail#184] [General]
I have SMTP service with authentication of GSSPI NTLM, is there anyway to achieve using this library? I noticed there is WithSMTPAuthCustom function, I wonder if I can use this
![image](https://github.com/wneessen/go-mail/assets/58017453/1746d98e-5306-45dd-a5f1-b45e9eff57d0)

**Comment by wneessen** [comment]
Hi @ZinedineR,

we currently do not support GSSAPI for NTLM but you are right, `WithSMTPAuthCustom()` would be the way to do it in your local code. As long as your provided custom fuction, that implements the GSSAPI authentication for NTLM satisfies the `smtp.Auth` interface: https://pkg.go.dev/github.com/wneessen/go-mail@v0.4.1/smtp#Auth you should be able to work with it.

**Comment by ZinedineR** [comment]
so I have tried to copy ntlm.go from this library github.com/cmarkh/smtp, I have also changed so the Start(server *smtp.ServerInfo)  to your smtp function. Here's the problem:

their library: func (c *Client) Auth(a Auth)

```
case 334:
			if msg64 == "NTLM supported" {
				msg = []byte(msg64)
			} else {
				msg, err = encoding.DecodeString(msg64)
			}
```

your library: 

```
case 334:
				msg, err = encoding.DecodeString(msg64)
			
```

it seems because the return of msg64 is "NTLM supported", your smtp can decode the string properly. If you may have more free time, I would like to add NTLM auth to your library. I want to discuss more about this

**Reply by wneessen** [reply]
PRs are always welcome. When porting code from other libraries, though please make sure that their license allows it. Also please keep in mind that one of go-mail's principles is to only rely on the Go stdlib.

**Comment by ZinedineR** [comment]
Allright, thanks wneessen, I'll research more about it.

**Comment by mkalus** [comment]
Maybe a bit late, but I have created a small custom handler that should do the trick:

```go
import (
	"fmt"
	"github.com/songxiang93/go-ntlm/ntlm"
	"github.com/wneessen/go-mail/smtp"
)

type NTLMCustomAuth struct {
	UserName string
	Password string
	Domain   string
	session  ntlm.ClientSession
}

func (a *NTLMCustomAuth) Start(info *smtp.ServerInfo) (proto string, toServer []byte, err error) {
	if len(info.Auth) > 0 {
		supported := false
		for _, method := range info.Auth {
			if method == "NTLM" {
				supported = true
			}
		}
		if !supported {
			return "", nil, fmt.Errorf("NTLM auth method not supported")
		}

		a.session, err = ntlm.CreateClientSession(ntlm.Version2, ntlm.ConnectionlessMode)
		if err != nil {
			return "", nil, fmt.Errorf("failed to create NTLM client session: %v", err)
		}
		a.session.SetUserInfo(a.UserName, a.Password, a.Domain)
		negotiate, err := a.session.GenerateNegotiateMessage()
		if err != nil {
			return "", nil, fmt.Errorf("failed to generate negotiate message: %v", err)
		}
		return "NTLM", negotiate.Bytes(), nil
	}
	return "NTLM", nil, nil
}

func (a *NTLMCustomAuth) Next(challengeBytes []byte, more bool) (toServer []byte, err error) {
	if more {
		challenge, err := ntlm.ParseChallengeMessage(challengeBytes)
		if err != nil {
			return nil, fmt.Errorf("failed to parse challenge message: %v", err)
		}
		err = a.session.ProcessChallengeMessage(challenge)
		if err != nil {
			return nil, fmt.Errorf("failed to process challenge message: %v", err)
		}

		authenticate, err := a.session.GenerateAuthenticateMessage()
		if err != nil {
			return nil, fmt.Errorf("failed to generate authenticate message: %v", err)
		}
		return authenticate.Bytes(), nil
	}
	return nil, nil
}
```

**Reply by mkalus** [reply]
Ah yes, and I used github.com/songxiang93/go-ntlm/ntlm for the ntlm library, it was a bit of a search to get one that did the trick.

**Reply by wneessen** [reply]
Thanks @mkalus! I'll reference this in the Wiki in case somebody else needs this at some point.

**Reply by wneessen** [reply]
Wiki updated: https://github.com/wneessen/go-mail/wiki/NTLM-Auth
